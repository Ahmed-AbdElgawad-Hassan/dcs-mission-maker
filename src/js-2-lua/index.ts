export * from "./js-2-lua";
