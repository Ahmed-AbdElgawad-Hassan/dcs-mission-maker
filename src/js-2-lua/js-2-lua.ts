import { forOwn, isArray, isEmpty, isObject, isString, range } from "lodash";

export const js2Lua = (data: any, depth = 0): string => {
  const indentation = range(0, depth + 1)
    .map(() => "")
    .join("\t");

  if (isArray(data)) {
    if (isEmpty(data)) {
      return `\n${indentation}{\n${indentation}}`;
    }
    return `\n${indentation}{\n${data
      .map((it, idx) => `${indentation}\t[${idx + 1}]=${js2Lua(it, depth + 1)}`)
      .join(",\n")}\n${indentation}}`;
  }

  if (isObject(data)) {
    if (isEmpty(data)) {
      return `\n${indentation}{\n${indentation}}`;
    }
    let str: string[] = [];
    // Don't use loadsh map as some objects can have "length: 0"
    forOwn(data, (value, key) =>
      str.push(`${indentation}\t[${js2Lua(key)}]=${js2Lua(value, depth + 1)}`),
    );
    return `\n${indentation}{\n${str.join(",\n")}\n${indentation}}`;
  }

  if (typeof data === "number") {
    return data.toString();
  }

  if (isString(data) && data.match(/^_\d+/)) {
    return data.replace(/^_/, "")
  }

  if (isString(data)) {
    return JSON.stringify(data);
  }

  return `${data}`;
};
