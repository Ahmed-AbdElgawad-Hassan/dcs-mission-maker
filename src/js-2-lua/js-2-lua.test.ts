import { js2Lua } from "./js-2-lua";

describe("js2Lua", () => {
  it("should convert objects to key with square brackets", () => {
    expect(js2Lua({ x: -291014, y: 617414 })).toMatchSnapshot();
    expect(js2Lua({ id: 21, name: "Australia" })).toMatchSnapshot();
  });

  it("should convert array to indexed keys with square brackets", () => {
    expect(
      js2Lua([
        { id: 21, name: "Australia" },
        { id: 15, name: "Israel" },
      ]),
    ).toMatchSnapshot();
  });
});
