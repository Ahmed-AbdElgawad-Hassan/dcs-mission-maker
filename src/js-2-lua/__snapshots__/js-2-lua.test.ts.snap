// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`js2Lua should convert array to indexed keys with square brackets 1`] = `
"
{
	[1]=
	{
		["id"]=21,
		["name"]="Australia"
	},
	[2]=
	{
		["id"]=15,
		["name"]="Israel"
	}
}"
`;

exports[`js2Lua should convert object keys to int where number is prefixed with _ 1`] = `
"
{
	[1]=1,
	[2]=1,
	[3]=1,
	["name"]="Enfield11",
	["_ignoreme"]="hello"
}"
`;

exports[`js2Lua should convert objects to key with square brackets 1`] = `
"
{
	["x"]=-291014,
	["y"]=617414
}"
`;

exports[`js2Lua should convert objects to key with square brackets 2`] = `
"
{
	["id"]=21,
	["name"]="Australia",
	["luaScript"]="-- This is a script you can put in a mission\\n\\n-- Limits that encompass the entire green area in the mission editor\\nlocal minY = -600000\\nlocal maxY = 1200000\\nlocal minX = -700000\\nlocal maxX = 400000\\n-- Data is sampled every 50km\\nlocal step = 50000\\n\\nfor xCoord = minX, maxX, step do\\n local row = {}\\n local rowString = \\"{\\"\\n for yCoord = minY, maxY, step do\\n local convLat, convLon = coord.LOtoLL({x = xCoord, y = 0, z = yCoord})\\n local lookupPoint = {{yCoord, xCoord}, {convLat, convLon}}\\n rowString = rowString .. string.format(\\"{{%f,%f},{%f,%f}},\\", yCoord, xCoord, convLat, convLon)\\n end\\n rowString = rowString .. \\"},\\\\n\\"\\n env.info(rowString)\\nend"
}"
`;

exports[`js2Lua should convert objects with length property 1`] = `
"
{
	["x"]=-291014,
	["y"]=617414,
	["length"]=0
}"
`;
